rs/replicated_state/src/metadata_state.rs

```diff
@@ -760,6 +765,8 @@ impl TryFrom<(pb_metadata::SystemMetadata, &dyn CheckpointLoadingMetrics)> for S
                 Some(blockmaker_metrics) => (blockmaker_metrics, metrics).try_into()?,
                 None => BlockmakerMetricsTimeSeries::default(),
             },
+            // TODO: Review. This could be incorrect. Would that have consequences?
+            replica_version: ReplicaVersion::default(),
```

Comment:
This is where the deserialisation happens after checkpointing. That should definitely not be the default, but `item.replica_version`. 

But there is also more to think about here. You have the problem that in `pb_metadata::SystemMetadata` the field is an `Option` but in `SystemMetadata` it is not. 

You need to ensure two properties:
1. Serialize and then deserialize should always give you the same value back. This you should get if you keep serialize like now (as in always `Some(...)`), and change deserialise to use `item.replica_version` if it is not None.
2. During replica upgrades the first time we load the state the value of `item.replica_version` is going to be `None`. So we have to make sure that we handle this case correctly. Without understanding your full code yet, it's probably good enough to set it to `default()` here and have it being overwritten when executing the first batch.

Reply:
The reasons why you possibly saw errors during subnet splitting in particular is that there we load a checkpoint and then write it back without a block inbetween. So relying on the `None` case being overwritten won't work in general, and that special handling should be limited to the one time where we upgrade.

Reply:
Ah, it was an oversight that here I actually do have an `item` available. Thanks for the context.

Reply:
Alternatively, do you even need to persist the replica version in the state? Since it is delivered in every batch, you could imagine a solution where execution of the new management canister function simply returns whatever was delivered in the batch of the current round.

Reply:
Perhaps there is another way, but I am not sure how that would look exactly. If the data is in the replicated state, it is easy to reason about its validity, and it is easy to retrieve where it's needed.

Reply:
In favor of Leo's argument, I will point out that retrieving the replica version from the state might mean that e.g. after a replica upgrade a new replica will return an old replica version (because that's what's in the state).

I suppose one could take care to always update the replica version at the beginning of the round (as we do e.g. with `batch_time` and `network_topology`) so during replicated execution we always have an up-to-date version. But this would still be problematic for queries. And, taking a step back, it introduces an extra level of indirection that is not required, strictly speaking.

Reply:
If the replicated state is an indirection, can you help me find a more direct way? 

I know that I need the `replica_version` in [`execute_subnet_message`](https://sourcegraph.com/github.com/dfinity/ic/-/blob/rs/execution_environment/src/execution_environment.rs?L468). From there I have access to the replicated state, some (to me) unrelated arguments, and `self: ExecutionEnvironment`, which contains a bunch of fields which again, to me, are not related to `Block`s or `Batch`es. 

Conversely, all I can see [`deliver_batches`](https://sourcegraph.com/github.com/dfinity/ic/-/blob/rs/consensus/src/consensus/batch_delivery.rs?L46) and [`deliver_batch`](https://sourcegraph.com/github.com/dfinity/ic/-/blob/rs/messaging/src/message_routing.rs?L1426) do is interact with message routing and possibly the state manager... so how would you thread through the information if not via the replicated state?

Reply:
The `Batch` is taken apart [here](https://sourcegraph.com/github.com/dfinity/ic/-/blob/rs/messaging/src/state_machine.rs?L163) and then execution is triggered.

Not sure if that is easier and/or better than putting it into the ReplicatedState though.

Reply:
I still need to store it somewhere, not sure what a better place would be. `StateManagerImpl`? Then I'd also have to add a trait method to retrieve the replica_version from it. That does not seem right.

Reply:
You can put it into the [RegistryExecutionSettings](https://sourcegraph.com/github.com/dfinity/ic/-/blob/rs/interfaces/src/execution_environment.rs?L1182:12-1182:37) [here](https://sourcegraph.com/github.com/dfinity/ic@0b29cd1e41ce09bae0d98ccedb4ce9de8532e0e6/-/blob/rs/messaging/src/message_routing.rs?L797) and then you have that available in [execute_subnet_message](https://sourcegraph.com/github.com/dfinity/ic/-/blob/rs/execution_environment/src/execution_environment.rs?L475).

Reply:
I cannot use the version from the registry, I need the version from the most recent block/batch. 

The reason is that outside entities can always learn the replica version the registry prescribes for a given subnet, but they currently cannot learn the actually running version. There is a discrepancy between the two during the replica upgrade process, and I need to know when the upgrade has succeeded - that is when the registry's version and the replica's version agree.

Reply:
@eichhorl do you see a way to connect the dots in the way you described? I'm afraid I still don't.

Reply:
> I suppose one could take care to always update the replica version at the beginning of the round (as we do e.g. with batch_time and network_topology) so during replicated execution we always have an up-to-date version. But this would still be problematic for queries.

IIUC (?) the value would only be returned by a management canister call, so queries shouldn't be a problem.

> @eichhorl do you see a way to connect the dots in the way you described? I'm afraid I still don't.

I think what @schneiderstefan meant is that you could add a new field to the `RegistryExecutionSettings`, that would then hold the replica version extracted from the batch. 

More generally, we do something similar for the `idkg_subnet_public_keys` for instance. Those are taken from the batch [here](https://sourcegraph.com/github.com/dfinity/ic/-/blob/rs/messaging/src/state_machine.rs?L166), and then passed to execution, where they are returned from a management canister call [here](https://sourcegraph.com/github.com/dfinity/ic/-/blob/rs/execution_environment/src/execution_environment.rs?L1027), without going through the replicated state.

Reply:
PTAL:
https://github.com/dfinity/ic/pull/2201

If 2201 is preferred, I will close this PR.